tools/codegen/dest/ufunc.py

```diff
@@ -0,0 +1,440 @@
+        ctor_tensor = 'other',
+        ufunc_key = UfuncKey.CUDAFunctorOnSelf,
+    ),
+]
+
+def compute_ufunc_cuda_dtype_body(g: NativeFunctionsGroup, dtype: ScalarType, inner_loops: Dict[UfuncKey, UfunctorSignature], parent_ctx: Sequence[Binding]) -> str:
+    body = "using opmath_t = at::opmath_type<scalar_t>;"
+    body += "if (false) {}\n";
+    for config in BinaryScalarSpecializationConfigs:
+        if config.ufunc_key not in inner_loops:
+            continue
+        ufunctor_sig = inner_loops[config.ufunc_key]
+        scalar_idx = config.scalar_idx + 1
+        # Make a copy and at the same time widen the type (not permissible
+        # without copy; we don't want to mutate the input argument anyway)
+        ctx: List[Union[Expr, Binding]] = [b for b in parent_ctx]
+        ctx.append(Expr(
+            expr=f"iter.scalar_value<opmath_t>({scalar_idx})",
+            type=NamedCType(config.ctor_tensor, BaseCType(opmath_t)),
+        ))
+        ufunctor_ctor_exprs_str = ', '.join(a.expr for a in translate(ctx, ufunctor_sig.arguments().ctor))
+
+        # NB: ufunctor must be allocated before iter.remove_operand is called,
+        # as it relies on iter
+        body += f"""\
+else if (iter.is_cpu_scalar({scalar_idx})) {{
+  {ufunctor_sig.name}<scalar_t> ufunctor({ufunctor_ctor_exprs_str});
+  iter.remove_operand({scalar_idx});
+  gpu_kernel(iter, ufunctor);
+}}"""
+
+    ufunctor_sig = inner_loops[UfuncKey.CUDAFunctor]
+    ufunctor_ctor_exprs_str = ', '.join(a.expr for a in translate(ctx, ufunctor_sig.arguments().ctor))
+    body += f"""
+else {{
+  gpu_kernel(iter, {ufunctor_sig.name}<scalar_t>({ufunctor_ctor_exprs_str}));
+}}
+    """
+    return body
+
+@with_native_function
+def compute_ufunc_cuda(g: NativeFunctionsGroup) -> str:
+    # First, build the functors, indexing them by dtype
+    ufunctor_sigs, ufunctors = compute_ufunc_cuda_functors(g)
+
+    # Next, build the conditionals
+    sig = StructuredImplSignature(g, ufunc.kernel_name(g, DispatchKey.CUDA))
+    dtype_cases = []
+    for dtype, inner_ufunctor_sigs in ufunctor_sigs.items():
+        dtype_cases.append(f"""
+AT_PRIVATE_CASE_TYPE("{sig.name}", at::ScalarType::{dtype}, {ScalarTypeToCppMapping[dtype]},
+  [&]() {{
+    {compute_ufunc_cuda_dtype_body(g, dtype, inner_ufunctor_sigs, sig.arguments())}
+  }}
+)
+""")
+
+    dtype_cases_str = "\n".join(dtype_cases)
+
+    return f"""
+{ufunctors}
+
+{sig.defn()} {{
+  TensorIteratorBase& iter = *this;
+  at::ScalarType st = iter.common_dtype();
+  RECORD_KERNEL_FUNCTION_DTYPE("{sig.name}", st);
+  switch (st) {{
+    {dtype_cases_str}
+    default:
+      TORCH_CHECK(false, "{sig.name}", " not implemented for '", toString(st), "'");
+  }}
+}}
+"""
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+#
+#                                   CPU STUFF
+#
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+
+@dataclass(frozen=True)
+class StubSignature:
```

Comment:
Is there a reason why CUDA doesn't share the `DispatchStub` with CPU kernels? The `TensorItensorBase &` signature of the dispatch stub is preferable because you don't need to include `Tensor.h` and so kernels aren't recompiled when a method operator's signature changes.

Reply:
My reasoning was that DispatchStub is unnecessary for CUDA operators, since once the dispatcher has put you in native/cuda, there is no need for a second dispatch. I wasn't thinking about recompilation avoidance though!

However, I'm not sure Tensor.h is actually getting included here? Here are the includes I end up with in this file:

```
#include <ATen/native/ufunc/add.h>
#include <ATen/Dispatch.h>
#include <c10/core/Scalar.h>
#include <ATen/native/cuda/Loops.cuh>
```

This seems, at least to me, to be as good as it gets!

Reply:
If that's true then great! Please add `#define TORCH_ASSERT_NO_OPERATORS` to the template to enforce that going into the future.

I was expecting Tensor be included for building the `TensorIteratorBase` but I see that this file doesn't actually build it because the meta function is inherited from elsewhere. That should be fine so long as the function signatures can be declared without including tensor.